### PR TITLE
Home-away scheduling fixes

### PR DESCRIPTION
**Pull request by severinson**
Tolerations added via home-away scheduling are not passed through correctly. This PR fixes the issue by:
1. Setting max priority correctly when scheduling away jobs.
2. Passing through tolerations on evicted jobs to the next scheduling attempt.

There are unit tests confirming the issue is solved. I've also tested that it works end-to-end manually.

### CODE REVIEW

File: internal/scheduler/gang_scheduler_test.go [hunk 1]
```diff
@@ -513,9 +513,56 @@ func TestGangScheduler(t *testing.T) {
...
 				return
 			}(),
-			ExpectedScheduledIndices:       []int{1},
-			ExpectedScheduledJobs:          []int{0, 1},
-			ExpectedRuntimeGangCardinality: []int{0, 1},
+			ExpectedScheduledIndices:        []int{1},
+			ExpectedCumulativeScheduledJobs: []int{0, 1},
+			ExpectedRuntimeGangCardinality:  []int{0, 1},
+		},
+		"Home-away scheduling": {
```
Comment (severinson) [inline internal/scheduler/gang_scheduler_test.go:520]: This tests that the priority is set correctly. Passes since adding the priority fix in nodedb.go.

File: internal/scheduler/nodedb/nodedb.go [hunk 2]
```diff
@@ -768,7 +761,9 @@ func (nodeDb *NodeDb) selectNodeForJobWithUrgencyPreemption(
 			continue
 		}
 
-		if priority > req.Priority {
+		// Using pctx.ScheduledAtPriority instead of jctx.PodRequirements.Priority,
+		// since the pctx.ScheduledAtPriority may differ, e.g., in case of home-away scheduling.
+		if priority > pctx.ScheduledAtPriority {
```
Comment (severinson) [inline internal/scheduler/nodedb/nodedb.go:766]: This was the actual issue to be fixed in this file. Other changes in this file is just refactoring for clarity.

File: internal/scheduler/preempting_queue_scheduler.go [hunk 3]
```diff
@@ -852,12 +853,20 @@ func (evi *Evictor) Evict(ctx *armadacontext.Context, nodeDbTxn *memdb.Txn) (*Ev
...
+			// 2. Add a node selector ensuring the job can only be re-scheduled onto the node it was evicted from.
+			// 3. Add tolerations for all taints on the node. This to ensure that:
+			//    - Adding taints to a node doesn't cause jobs already running on the node to be preempted.
+			//    - Jobs scheduled as away jobs have the necessary tolerations to be re-scheduled.
+			// TODO(albin): We can remove the checkOnlyDynamicRequirements flag in the nodeDb now that we've added the tolerations.
 			jctx := schedulercontext.JobSchedulingContextFromJob(evi.priorityClasses, job)
 			jctx.IsEvicted = true
 			jctx.AddNodeSelector(schedulerconfig.NodeIdLabel, node.Id)
 			evictedJctxsByJobId[job.GetId()] = jctx
+			for _, taint := range node.Taints {
```
Comment (severinson) [inline internal/scheduler/preempting_queue_scheduler.go:867]: This was the second issue. We didn't previously pass through tolerations added when scheduling a job as an away-job.

File: internal/scheduler/preempting_queue_scheduler_test.go [hunk 4]
```diff
@@ -1938,6 +1938,16 @@ func TestPreemptingQueueScheduler(t *testing.T) {
 					assert.True(t, ok)
 					assert.NotEmpty(t, nodeId)
 
+					node, err := nodeDb.GetNode(nodeId)
+					require.NoError(t, err)
+					assert.NotEmpty(t, node)
+
+					// Check that the job can actually go onto this node.
+					matches, reason, err := nodedb.StaticJobRequirementsMet(node.Taints, node.Labels, node.TotalResources, jctx)
```
Comment (severinson) [inline internal/scheduler/preempting_queue_scheduler_test.go:1946]: Adding this caused an existing test to fail. Copying over the tolerations when evicting made that test pass.

File: pkg/armadaevents/events.proto [hunk 5]
```diff
@@ -6,10 +6,10 @@ option csharp_namespace = "ArmadaProject.Io.ArmadaEvents";
 import "k8s.io/api/core/v1/generated.proto";
 import "k8s.io/api/networking/v1/generated.proto";
 import "google/protobuf/timestamp.proto";
-import "google/protobuf/wrappers.proto";
```
Comment (severinson) [inline pkg/armadaevents/events.proto:9]: Drive-by change to remove a warning.

File: pkg/armadaevents/events.pb.go [hunk 6]
```diff
@@ -3526,233 +3526,233 @@ func init() {
 func init() { proto.RegisterFile("pkg/armadaevents/events.proto", fileDescriptor_6aab92ca59e015f8) }
 
 var fileDescriptor_6aab92ca59e015f8 = []byte{
-	// 3615 bytes of a gzipped FileDescriptorProto
```
Comment (severinson) [inline pkg/armadaevents/events.pb.go:3529]: 11 bytes saved!

**Review by zuqq — APPROVED** [review]
